### PR TITLE
feat(idtoken): add support for impersonated_service_account creds type

### PR DESCRIPTION
**Pull request by adrianajg**
- Allow NewTokenSource function to accept authorization credentials of type `impersonated_service_account`

Updates: #873

### CODE REVIEW

File: idtoken/idtoken.go [hunk 1]
```diff
@@ -67,6 +71,7 @@ func NewClient(ctx context.Context, audience string, opts ...ClientOption) (*htt
 // provided and configured with the supplied options. The parameter audience may
 // not be empty.
 func NewTokenSource(ctx context.Context, audience string, opts ...ClientOption) (oauth2.TokenSource, error) {
+	option.WithScopes()
```
Comment (codyoss) [inline idtoken/idtoken.go:74]: nit: this looks unintentional.

File: idtoken/idtoken.go [hunk 2]
```diff
@@ -103,45 +108,72 @@ func newTokenSource(ctx context.Context, audience string, ds *internal.DialSetti
...
+		if err != nil {
+			log.Println(err)
+		}
+		return ts, nil
 	}
-	return oauth2.ReuseTokenSource(tok, ts), nil
 }
 
-func isServiceAccount(data []byte) error {
+// isOfAllowedType returns the credentials type as a string, and an error.
```
Comment (codyoss) [inline idtoken/idtoken.go:160]: nit: update docs to reflect new name

File: idtoken/idtoken.go [hunk 3]
```diff
@@ -103,45 +108,72 @@ func newTokenSource(ctx context.Context, audience string, ds *internal.DialSetti
...
+// allowed types are "service_account" and "impersonated_service_account"
+func getAllowedType(data []byte) (string, error) {
```
Comment (codyoss) [inline idtoken/idtoken.go:162]: I wonder if instead this should return an enum(Go does not have such a thing but their are patterns). I think that would clean up this code a bit.

File: idtoken/idtoken.go [hunk 4]
```diff
@@ -103,45 +108,72 @@ func newTokenSource(ctx context.Context, audience string, ds *internal.DialSetti
...
 	}
+	// if not service account return an error
 	if err := json.Unmarshal(data, &f); err != nil {
-		return err
+		return "", err
 	}
-	if f.Type != "service_account" {
-		return fmt.Errorf("idtoken: credential must be service_account, found %q", f.Type)
+	if f.Type != "service_account" && f.Type != "impersonated_service_account" {
+		return "", fmt.Errorf("idtoken: credential must be service_account or impersonated_service_account, found %q", f.Type)
```
Comment (codyoss) [inline idtoken/idtoken.go:174]: More types may be added over time, be less specific and maybe just say it is an unsupported type.

File: idtoken/idtoken.go (hunk 3 above)
Comment (codyoss) [reply idtoken/idtoken.go:162]: If we did an enum I think this should not return an error, and then in the calling code handle the undefined enum state.

File: idtoken/idtoken.go [hunk 5]
```diff
@@ -103,45 +108,72 @@ func newTokenSource(ctx context.Context, audience string, ds *internal.DialSetti
...
-	customClaims["target_audience"] = audience
-
-	cfg.PrivateClaims = customClaims
-	cfg.UseIDToken = true
-
-	ts := cfg.TokenSource(ctx)
-	tok, err := ts.Token()
-	if err != nil {
-		return nil, err
+	if allowedType == "service_account" {
```
Comment (codyoss) [inline idtoken/idtoken.go:115]: Feels like this should we a switch instead of an if/else

File: idtoken/idtoken_test.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
+// Copyright 2020 Google LLC.
```
Comment (codyoss) [inline idtoken/idtoken_test.go:1]: nit: update year

File: idtoken/idtoken_test.go [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package idtoken
```
Comment (codyoss) [inline idtoken/idtoken_test.go:5]: I think this test should either be removed for now or else we need to turn it into a real integration test and setup creds. As is this is causing CI to fail.

**Review by codyoss — CHANGES_REQUESTED** [review]

File: idtoken/idtoken.go (hunk 1 above)
Comment (adrianajg) [reply idtoken/idtoken.go:74]: Removed.

**Comment by codyoss** [comment]
@adrianajg It appears you have files that are not properly formatted

File: idtoken/integration_test.go [hunk 8]
```diff
@@ -49,6 +50,58 @@ func TestNewTokenSource(t *testing.T) {
 	}
 }
 
+func TestNewTokenSource_WithImpersonatedServiceAccountCreds(t *testing.T) {
```
Comment (codyoss) [inline idtoken/integration_test.go:53]: Let remove this for now.

Comment (JohnsoNeal2013) [reply idtoken/integration_test.go]: Yea

**Comment by JohnsoNeal2013** [comment]
Can someone please help me

**Review by adrianajg — COMMENTED** [review]
Addressed changes, let me know if you see anything else. Thanks!

**Review by codyoss — APPROVED** [review]

**Pull request merged**